skyportal/models.py

```diff
@@ -1645,39 +1652,170 @@ class Spectrum(Base):
+           filesystem.
+        obj_id : str, optional
+           The id of the Obj that this Spectrum is of, if not present
+           in the ASCII header.
+        instrument_id : int, optional
+           ID of the Instrument with which this Spectrum was acquired,
+           if not present in the ASCII header.
+        observed_at : string or datetime, optional
+           Median UTC ISO time stamp of the exposure or exposures in which
+           the Spectrum was acquired, if not present in the ASCII header."
+        wave_colindex: integer, optional
+           The 0-based index of the ASCII column corresponding to the wavelength
+           values of the spectrum (default 0).
+        flux_colindex: integer, optional
+           The 0-based index of the ASCII column corresponding to the flux
+           values of the spectrum (default 1).
+        fluxerr_colindex: integer or None, optional
+           The 0-based index of the ASCII column corresponding to the flux error
+           values of the spectrum (default 2). If there are only 2 columns
+           in the input file this value will be ignored. If there are more than
+           2 columns in the input file, but none of them correspond to flux error
+           values, set this parameter to None.
 
         Returns
         -------
-
         spec : `skyportal.models.Spectrum`
            The Spectrum generated from the ASCII file.
 
         """
-        data = np.loadtxt(filename)
-        if data.shape[1] != 2:  # TODO support other formats
-            raise ValueError(f"Expected 2 columns, got {data.shape[1]}")
+
+        filename = Path(filename)
+
+        if data is None:
+            try:
+                table = ascii.read(filename, comment='#', header_start=None)
+            except Exception as e:
+                e.args = (f'Error parsing ASCII file: {e.args[0]}',)
+                raise
+            bytes = open(filename).read().encode('ascii')
+        else:
+            file_obj = io.BytesIO(data.encode('ascii'))
+            try:
+                table = ascii.read(file_obj, comment='#', header_start=None)
+            except Exception as e:
+                e.args = (f'Error parsing ASCII file: {e.args[0]}',)
+                raise
+            bytes = data
+
+        tabledata = np.asarray(table)
+        colnames = table.colnames
+
+        if len(colnames) < 2:
+            raise ValueError(
+                'Input data must have at least 2 columns (wavelength, '
+                'flux, and optionally flux error).'
+            )
+
+        if 'comments' in table.meta:
+
+            # this section matches lines like:
+            # XTENSION: IMAGE
+            # BITPIX: -32
+            # NAXIS: 2
+            # NAXIS1: 433
+            # NAXIS2: 1
+
+            header = {}
+            for line in table.meta['comments']:
+                try:
+                    result = yaml.load(line, Loader=yaml.FullLoader)
```

Comment:
This can't be very effective?  Why not bundle and load all at once?

Reply:
The reason I do it line by line is because, if we parse the whole header as yaml and a single comment line is not formatted as yaml but is instead of a different form, e.g., the third row here:

``` 
# key1: value1
# key2: value2
# col1 col2 col3 col4 
```

then parsing of the entire header will fail. Having one or two generic comment lines in the header is common: they often contain column labels or some similar data that we don't need to keep track of or use in any way. Similar to how we don't want an improperly configured cosmology to cause the entire app to crash, we don't want a single misformatted header line to cause the entire upload to fail. 

On the frontend, we will use the `SpectrumParser` handler to return the parsed ascii spectrum for users to inspect before they can click an upload button to actually post the spectrum to the database.